### PR TITLE
Use os_sys_calls_impl for setting TCP socket options

### PR DESCRIPTION
**Pull request by cmluciano**
Signed-off-by: Christopher M. Luciano <cmluciano@us.ibm.com>

*Description*: Addresses TODO to use os_sys_calls_impl for setting TCP socket options
*Risk Level*: Low
*Testing*: `bazel test //test/common/network:listen_socket_impl_test`

### CODE REVIEW

**Comment by htuch** [comment]
@cmluciano needs fix_format.

File: source/common/network/listen_socket_impl.cc [hunk 1]
```diff
@@ -42,10 +43,11 @@ TcpListenSocket::TcpListenSocket(const Address::InstanceConstSharedPtr& address,
...
   RELEASE_ASSERT(fd_ != -1, "");
 
-  // TODO(htuch): This might benefit from moving to SocketOptionImpl.
   int on = 1;
-  int rc = setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
-  RELEASE_ASSERT(rc != -1, "");
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  Api::SysCallIntResult status =
+      os_syscalls.setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+  RELEASE_ASSERT(status.rc_ != -1, fmt::format("failed to set SO_REUSEADDR socket option"));
```
Comment (venilnoronha) [inline source/common/network/listen_socket_impl.cc]: We can get rid of the `fmt::format` here.

**Comment by cmluciano** [comment]
@htuch Had to force-push to get DCO to work, but this is now cleanly rebased.

File: source/common/network/listen_socket_impl.cc [hunk 2]
```diff
@@ -48,10 +48,12 @@ void ListenSocketImpl::setupSocket(const Network::Socket::OptionsSharedPtr& opti
 template <>
 void NetworkListenSocket<
     NetworkSocketTrait<Address::SocketType::Stream>>::setPrebindSocketOptions() {
-  // TODO(htuch): This might benefit from moving to SocketOptionImpl.
+
   int on = 1;
-  int rc = setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
-  RELEASE_ASSERT(rc != -1, "");
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
```
Comment (jmarantz) [inline source/common/network/listen_socket_impl.cc]: drive-by comment: Can we plumb the Api::Api& through to here or do we need to depend on a static?

Comment (jmarantz) [reply source/common/network/listen_socket_impl.cc]: ...I just noticed that this was already approved by @htuch so feel free to ignore for now...but maybe this could be a follow-up.

Comment (htuch) [reply source/common/network/listen_socket_impl.cc]: @jmarantz historically we were cool with this singleton for OS syscalls, since it seemed a pain to plumb Api everywhere just to get a libc mock capability. Have we decided to phase that out?

**Comment by cmluciano** [comment]
@jmarantz @htuch Let me know if you want me to include the changes above.
